File: solutions/240.py

def main():
    import sys
    data = sys.stdin.read().splitlines()
    if not data:
        print(0)
        return
        
    n = int(data[0].strip())
    views = []
    index = 1
    for _ in range(6):
        view = []
        for i in range(n):
            line = data[index].strip()
            index += 1
            view.append(line)
        views.append(view)
    
    front = views[0]
    left = views[1]
    back = views[2]
    right = views[3]
    top = views[4]
    bottom = views[5]
    
    cube = [[[True for _ in range(n)] for _ in range(n)] for _ in range(n)]
    
    for x in range(n):
        for y in range(n):
            for z in range(n):
                if front[z][x] == '.':
                    if cube[x][y][z]:
                        cube[x][y][z] = False
                if left[z][y] == '.':
                    if cube[x][y][z]:
                        cube[x][y][z] = False
                if back[z][n-1-x] == '.':
                    if cube[x][y][z]:
                        cube[x][y][z] = False
                if right[z][n-1-y] == '.':
                    if cube[x][y][z]:
                        cube[x][y][z] = False
                if top[n-1-y][x] == '.':
                    if cube[x][y][z]:
                        cube[x][y][z] = False
                if bottom[y][x] == '.':
                    if cube[x][y][z]:
                        cube[x][y][z] = False
    
    for x in range(n):
        for z in range(n):
            color = front[z][x]
            if color != '.':
                found = False
                for y in range(n):
                    if cube[x][y][z] and (front[z][x] == color or front[z][x] != '.'):
                        found = True
                        break
                if not found:
                    for y in range(n):
                        if cube[x][y][z]:
                            cube[x][y][z] = False
    
    for y in range(n):
        for z in range(n):
            color = left[z][y]
            if color != '.':
                found = False
                for x in range(n):
                    if cube[x][y][z] and (left[z][y] == color or left[z][y] != '.'):
                        found = True
                        break
                if not found:
                    for x in range(n):
                        if cube[x][y][z]:
                            cube[x][y][z] = False
    
    for x in range(n):
        for z in range(n):
            color = back[z][n-1-x]
            if color != '.':
                found = False
                for y in range(n):
                    if cube[x][y][z] and (back[z][n-1-x] == color or back[z][n-1-x] != '.'):
                        found = True
                        break
                if not found:
                    for y in range(n):
                        if cube[x][y][z]:
                            cube[x][y][z] = False
    
    for y in range(n):
        for z in range(n):
            color = right[z][n-1-y]
            if color != '.':
                found = False
                for x in range(n):
                    if cube[x][y][z] and (right[z][n-1-y] == color or right[z][n-1-y] != '.'):
                        found = True
                        break
                if not found:
                    for x in range(n):
                        if cube[x][y][z]:
                            cube[x][y][z] = False
    
    for x in range(n):
        for y in range(n):
            color = top[n-1-y][x]
            if color != '.':
                found = False
                for z in range(n):
                    if cube[x][y][z] and (top[n-1-y][x] == color or top[n-1-y][x] != '.'):
                        found = True
                        break
                if not found:
                    for z in range(n):
                        if cube[x][y][z]:
                            cube[x][y][z] = False
    
    for x in range(n):
        for y in range(n):
            color = bottom[y][x]
            if color != '.':
                found = False
                for z in range(n):
                    if cube[x][y][z] and (bottom[y][x] == color or bottom[y][x] != '.'):
                        found = True
                        break
                if not found:
                    for z in range(n):
                        if cube[x][y][z]:
                            cube[x][y][z] = False
    
    count = 0
    for x in range(n):
        for y in range(n):
            for z in range(n):
                if cube[x][y][z]:
                    count += 1
                    
    print(count)

if __name__ == "__main__":
    main()

File: solutions/240.go

import sys

def solve() -> None:
    data = sys.stdin.read().strip().splitlines()
    if not data:
        return
    N = int(data[0].strip())
    rows = [line.strip().split() for line in data[1:1 + N]]

    # pictures as N strings of length N
    front = [list(rows[i][0]) for i in range(N)]
    left =  [list(rows[i][1]) for i in range(N)]
    back =  [list(rows[i][2]) for i in range(N)]
    right = [list(rows[i][3]) for i in range(N)]
    top =   [list(rows[i][4]) for i in range(N)]
    bottom=[list(rows[i][5]) for i in range(N)]

    # emptiness flags for the three families of lines
    empty_z = [[False] * N for _ in range(N)]   # front / back direction
    for y in range(N):
        for x in range(N):
            i = N - 1 - y          # row index in front/back pictures
            j = x                  # column index
            if front[i][j] == '.' or back[i][j] == '.':
                empty_z[x][y] = True

    empty_x = [[False] * N for _ in range(N)]   # left / right direction
    for y in range(N):
        for z in range(N):
            i = N - 1 - y          # row index in left/right pictures
            j = z                  # column index
            if left[i][j] == '.' or right[i][j] == '.':
                empty_x[y][z] = True

    empty_y = [[False] * N for _ in range(N)]   # top / bottom direction
    for x in range(N):
        for z in range(N):
            i = z                  # row index in bottom picture
            j = x                  # column index
            if bottom[i][j] == '.' or top[i][j] == '.':
                empty_y[x][z] = True

    forced = 0
    for x in range(N):
        for y in range(N):
            for z in range(N):
                if empty_z[x][y] or empty_x[y][z] or empty_y[x][z]:
                    forced += 1

    answer = N * N * N - forced
    print(answer)


if __name__ == '__main__':
    solve()
